File: src_local/agents/Python/Audit/testify.py
from src_local.agents.base import BaseActivePersona
import logging
import time
from pathlib import Path

logger = logging.getLogger(__name__)

class TestifyPersona(BaseActivePersona):
    """
    Core: PhD in Software Verification & Reliability Strategy 🧪
    Estrategista que interpreta dados de verificação delegados a assistentes.
    """
    
    def __init__(self, project_root):
        super().__init__(project_root)
        self.name, self.emoji, self.role, self.stack = "Testify", "🧪", "PhD QA Strategist", "Python"
        # Agentes de Suporte ao QA
        from src_local.agents.Support.test_runner import TestRunner
        from src_local.agents.Support.quality_analyst import QualityAnalyst
        from src_local.agents.Support.pyramid_analyst import PyramidAnalyst
        self.runner = TestRunner()
        self.analyst = QualityAnalyst()
        self.pyramid_analyst = PyramidAnalyst()

    def perform_audit(self) -> list:
        start_time = time.time()
        logger.info(f"[{self.name}] Analisando Confiabilidade Estratégica...")
        
        audit_rules = [
            {'regex': 'def test_.*:\\s+pass', 'issue': 'Vazio: Teste sem asserções.', 'severity': 'critical'},
            {'regex': r"hypothesis", 'issue': 'Avançado: Teste de Propriedade detectado.', 'severity': 'low'}
        ]
        
        results = self.find_patterns(('.py', '.dart', '.kt'), audit_rules)
        self._log_performance(start_time, len(results))
        return results

    def run_test_suite(self):
        """DELEGAÇÃO: Executa testes via TestRunner e interpreta falhas."""
        logger.info("🧪 [Testify] Solicitando execução mecânica ao TestRunner...")
        metrics = self.runner.run_unittest_discover(self.project_root)
        
        # O estrategista adiciona a interpretação humana/PhD ao resultado bruto
        if not metrics["success"]:
            metrics["interpretation"] = self._interpret_failures(metrics.get("raw_output", ""))
            
        return metrics

    def _interpret_failures(self, raw_output):
        """Motor de Raciocínio sobre falhas (Visão PhD)."""
        if "ModuleNotFoundError" in raw_output: return "Quebra de dependência estrutural."
        if "AssertionError" in raw_output: return "Desvio de lógica funcional."
        return "Instabilidade técnica indeterminada."

    def analyze_test_quality_matrix(self, map_data=None):
        """DELEGAÇÃO: Calcula matriz de confiança via QualityAnalyst."""
        data = map_data if map_data is not None else self.context_data
        return self.analyst.calculate_confidence_matrix(data)

    def analyze_test_pyramid(self, map_data=None):
        """DELEGAÇÃO: Auditoria de distribuição delegada ao PyramidAnalyst."""
        data = map_data if map_data is not None else self.context_data
        return self.pyramid_analyst.analyze(data, self.read_project_file)

    def _reason_about_objective(self, objective, file, content):
        """Raciocínio sobre a profundidade da validação técnica."""
        info = self.context_data.get(file, {})
        if file.endswith('.py') and info.get("component_type") in ["AGENT", "CORE", "LOGIC", "UTIL"]:
            if not info.get("has_test", False):
                return f"Exposição de Risco: O objetivo '{objective}' exige confiança. O módulo '{file}' é Matéria Escura (Sem testes detectados)."
        
        # Detecção de testes vazios agora delegada ao AuditEngine via perform_audit
        return None

    def get_system_prompt(self):
        return f"Você é o Dr. {self.name}, estrategista de QA. Sua missão é garantir que o projeto siga a pirâmide de testes 70/20/10 e atinja 100% de confiança."

File: src_local/agents/Kotlin/Audit/testify.ts
import { BaseActivePersona } from "../../base.ts";
import type { AuditRule, StrategicFinding, AuditFinding } from "../../base.ts";
import type { ProjectContext } from "../../../core/types.ts";

/**
 * 🧪 Testify Persona (Kotlin Stack) - PhD in Quality Assurance
 */
export class TestifyPersona extends BaseActivePersona {
    constructor(projectRoot: string | undefined = undefined) {
        super(projectRoot);
        this.name = "Testify";
        this.emoji = "🧪";
        this.role = "PhD Quality Assurance Engineer";
        this.phd_identity = "Kotlin Test Coverage & QA";
        this.stack = "Kotlin";
    }

    override async execute(context: ProjectContext): Promise<any> {
        this.setContext(context);
        const findings = await this.performAudit();
        return findings;
    }

    override getAuditRules(): { extensions: string[]; rules: AuditRule[] } {
        return {
            extensions: [".kt"],
            rules: [
                { regex: /@Test\s+fun\s+\w+\(\)\s*{\s*}/, issue: "Teste Vazio: Função sem corpo detectada. Falsa cobertura PhD.", severity: "critical" },
                { regex: /!!\./, issue: "Fragilidade: Non-null assertion (!!) em teste. Pode causar falhas de runtime sem rastreabilidade PhD.", severity: "medium" },
                { regex: /try\s*{[\s\S]*?}\s*catch\s*\(e:\s*Exception\)\s*{\s*}/, issue: "Silenciamento: Catch de exceção vazio em teste. Oculta comportamentos anômalos.", severity: "high" }
            ]
        };
    }

    public override async performAudit(): Promise<any[]> {
        this.startMetrics();
        const rules = this.getAuditRules();
        const results = await this.findPatterns(rules.extensions, rules.rules);
        this.endMetrics(results.length);
        return results;
    }

    public Branding(): string { return `${this.emoji} ${this.name}`; }

    public override reasonAboutObjective(objective: string, file: string, content: string | Promise<string | null>): StrategicFinding | null {
        if (typeof content !== 'string') return null;
        if (content.includes('!!.')) {
            return {
                file, severity: "HIGH",
                issue: `Instabilidade de Teste: O objetivo '${objective}' exige segurança Kotlin. O uso de !! em '${file}' viola a integridade da 'Orquestração de IA'.`,
                context: "non-null assertion in test"
            };
        }
        return {
            file, severity: "INFO",
            issue: `PhD Quality (Kotlin): Analisando suíte para ${objective}. Focando em segurança de tipos e cobertura real.`,
            context: "analyzing quality"
        };
    }

    override selfDiagnostic(): any {
        return { status: "Soberano", score: 100, details: "Módulo de Qualidade Kotlin operando com rigor PhD." };
    }

    override getSystemPrompt(): string {
        return `Você é o Dr. Testify, PhD em qualidade de software e testes robustos para o ecossistema Kotlin/JVM.`;
    }
}
